src/rpc/blockchain.cpp

```diff
@@ -1645,34 +1646,141 @@ UniValue MempoolInfoToJSON(const CTxMemPool& pool)
     ret.pushKV("mempoolminfee", ValueFromAmount(std::max(pool.GetMinFee(maxmempool), ::minRelayTxFee).GetFeePerK()));
     ret.pushKV("minrelaytxfee", ValueFromAmount(::minRelayTxFee.GetFeePerK()));
     ret.pushKV("unbroadcastcount", uint64_t{pool.GetUnbroadcastTxs().size()});
+
+    if (feeLimits) {
+        const MempoolHistogramFeeRates& limits{feeLimits.value()};
+
+        /* Keep histogram per...
+         * ... cumulated tx sizes
+         * ... txns (count)
+         * ... cumulated fees */
+        std::vector<uint64_t> sizes(limits.size(), 0);
+        std::vector<uint64_t> count(limits.size(), 0);
+        std::vector<CAmount> fees(limits.size(), 0);
+
+        for (const CTxMemPoolEntry& e : pool.mapTx) {
+            const CAmount fee{e.GetFee()}, afees{e.GetModFeesWithAncestors()}, dfees{e.GetModFeesWithDescendants()};
+            const uint32_t size{(uint32_t)e.GetTxSize()}, asize{(uint32_t)e.GetSizeWithAncestors()}, dsize{(uint32_t)e.GetSizeWithDescendants()};
+
+            const CAmount fpb{CFeeRate{fee, size}.GetFee(1)};     // Fee rate per byte
+            const CAmount afpb{CFeeRate{afees, asize}.GetFee(1)}; // Fee rate per byte including ancestors
+            const CAmount dfpb{CFeeRate{dfees, dsize}.GetFee(1)}; // Fee rate per byte including descendants
```

Comment:
I think using ancestor feerates here is going to distort the result, because multiple transactions can have shared ancestors. Using descendant feerates is even more questionable to me, as they don't matter for transaction selection in blocks at all.

My suggestion would be either:
* (1) use only transaction feerates directly (no ancestors/descendants), and have a result that ignores CPFP-like mechanics. 
* (2) invoke the mining code to actually build a giant "block" from the entire mempool (ignoring weight limits), and have it report the feerates of its included packages (which are similar to what you'd get with ancestor fees/sizes, but is guaranteed not to count any transaction twice).

Reply:
> I think using ancestor feerates here is going to distort the result, because multiple transactions can have shared ancestors. Using descendant feerates is even more questionable to me, as they don't matter for transaction selection in blocks at all.

So I have adopted #15836 and did some modifications but I left L1665-L1667 as they were in the original PR because I was not sure why it was implemented that way. My attempts to contact the original author were unsuccessful for me to understand the motivation. Anyway, I knew back then that I didn't have a good explanation for the lines ... evidently rightly so. I should have added a comment warning about this though, that's my mistake, sorry for that.

Thank you for the analysis and the suggestions. I'll try to address the issue.

Reply:
@kiminuo Just in case that isn't clear: no need to apologize for trying to improve things. The final state of the code is a responsibility of everyone, developers and reviewers.

Reply:
@sipa

> * (1) use only transaction feerates directly (no ancestors/descendants), and have a result that ignores CPFP-like mechanics.

I have picked (1) approach for now because: I can modify the PR in reasonable time and then it can serve as some sort of baseline. So we can still decide later whether the approach is the best. 

Anyway, there are still questions: 

* Is (2) much better than (1) or not? 
* Is there room for having both (1) and (2) possibly?
* (2) seems hard-ish to implement. If that is true, one concern here is whether the additional complexity is worth it. I don't know.

btw: I was interested how https://jochen-hoenicke.de/queue/#BTC,24h,fee works and it seems that #15836 was actually heavily inspired by: https://github.com/jhoenicke/mempool/blob/31d8dcb896ebc8139d3e60c3f275b94811293777/mempool_sql.py#L44-L49.

Reply:
It really depends on the use case for the histogram. If people are trying to make decisions on what fee to use, it's probably better to under-estimate the fee rate of a given transaction (by applying a CPFP logic excessively) than over-estimate it (by ignoring CPFP, etc). Are there other use cases that would be impacted?

Reply:
> * Is (2) much better than (1) or not?

It depends on how much CPFP is going on of course in the mempool overall. I think (2) is "perfect" in the sense that it exactly answers the question how much vbytes are competing at every feerate level for block space, according to Bitcoin Core's own block building algorithm. But I think (1) is probably a good first approximation.

> * Is there room for having both (1) and (2) possibly?

I don't think so. We should just use the best algorithm we have, and I think (2) is strictly better, but obviously a lot more work.

> * (2) seems hard-ish to implement. If that is true, one concern here is whether the additional complexity is worth it. I don't know.

It's certainly more code changes. I'm not sure whether it's worth it, but that discussion can be left to a future improvement too.

Re: johoe's site using that formula... interesting find. This earlier version of the code has a bit more comments: https://github.com/jhoenicke/mempool/blob/548698d6d255a2e1b7d6f1981403d3de55c2182a/mempool-sql.pl#L30L38. I see where it's coming from now, but I can't imagine that it doesn't add as many inaccuracies as that it improves (see examples above).

Reply:
@luke-jr 

> It really depends on the use case for the histogram. If people are trying to make decisions on what fee to use, it's probably better to under-estimate the fee rate of a given transaction (by applying a CPFP logic excessively) than over-estimate it (by ignoring CPFP, etc).

From practicality point of view, I agree. But I would still like to follow [this plan](https://github.com/bitcoin/bitcoin/pull/21422#issuecomment-978941401) as it converges to the best possible solution.

@sipa 

> Re: johoe's site using that formula... interesting find. This earlier version of the code has a bit more comments: https://github.com/jhoenicke/mempool/blob/548698d6d255a2e1b7d6f1981403d3de55c2182a/mempool-sql.pl#L30L38. I see where it's coming from now, 

Ah, that's good to know. 

> but I can't imagine that it doesn't add as many inaccuracies as that it improves (see examples above).

Without thinking about it too deeply, my guess is that one can come up with pathological mempool state cases to force it report bad values. If it behaves ok-ish for real-world mempools, I don't really know. Anyway, https://github.com/bitcoin/bitcoin/pull/21422#issuecomment-978941401 makes sense to me (the worst alternative there is basically that people disagree and then we would chase approach (2) immediately).